### PR TITLE
Cap SVID lifetime by CA expiry

### PR DESCRIPTION
**Pull request by evan2645**
The agent renews SVIDs by inspecting their lifetime, however the CA
plugin will happily mint SVIDs that live past the lifetime of the CA
cert which signed them. As a result, given the right combinations of
TTLs, it is possible to miss a renewal that should have occurred.

This commit updates the CA plugin to truncate SVID lifetime if the
defined TTL exceeds that of the currently-active CA certificate.

Signed-off-by: Evan Gilman <evan@scytale.io>

### CODE REVIEW

File: pkg/server/plugin/ca/memory/memory_test.go [hunk 1]
```diff
@@ -221,7 +239,7 @@ func TestMemory_SignCsrExpire(t *testing.T) {
 	cert, err := x509.ParseCertificate(wcert.SignedCertificate)
 	roots := getRoots(t, m)
 	_, err = cert.Verify(x509.VerifyOptions{Roots: roots})
-	assert.Equal(t, "x509: certificate has expired or is not yet valid", err.Error())
+	assert.NotNil(t, err)
```
Comment (azdagron) [inline pkg/server/plugin/ca/memory/memory_test.go]: ignorable-nit: assert.NoError

File: pkg/server/plugin/ca/memory/memory_test.go [hunk 2]
```diff
@@ -231,7 +249,7 @@ func TestMemory_SignCsrNoCert(t *testing.T) {
 
 	wcert, err := m.SignCsr(ctx, &ca.SignCsrRequest{Csr: wcsr})
 
-	assert.Equal(t, "Invalid state: no certificate", err.Error())
+	assert.NotNil(t, err)
```
Comment (azdagron) [inline pkg/server/plugin/ca/memory/memory_test.go]: ignorable-nit: assert.NoError

File: pkg/server/plugin/ca/memory/memory_test.go [hunk 3]
```diff
@@ -242,17 +260,7 @@ func TestMemory_SignCsrErrorParsingSpiffeId(t *testing.T) {
...
-}
-
-func TestMemory_SignCsrErrorInvalidTTL(t *testing.T) {
-	m := populateCert(t)
-
-	wcsr := createWorkloadCSR(t, "spiffe://localhost")
-
-	wcert, err := m.SignCsr(ctx, &ca.SignCsrRequest{Csr: wcsr, Ttl: -5})
-	assert.Equal(t, "Invalid TTL: -5", err.Error())
+	assert.NotNil(t, err)
```
Comment (azdagron) [inline pkg/server/plugin/ca/memory/memory_test.go]: ignorable-nit: assert.NoError

File: pkg/server/plugin/ca/memory/memory.go [hunk 4]
```diff
@@ -221,51 +215,51 @@ func (m *MemoryPlugin) LoadCertificate(ctx context.Context, request *ca.LoadCert
 	}
 
 	if len(uris) != 1 {
-		return &ca.LoadCertificateResponse{}, fmt.Errorf("X.509 SVID certificates must have exactly one URI SAN. Found %v URI(s)", len(uris))
+		return &ca.LoadCertificateResponse{}, fmt.Errorf("load certificate: found %v URI(s); must have exactly one", len(uris))
```
Comment (azdagron) [inline pkg/server/plugin/ca/memory/memory.go]: one of these days we should discuss introducing something like github.com/zeebo/errs to the codebase to cut down on some of the verbosity around error creation (and get other benefits like stack traces).

File: pkg/server/plugin/ca/memory/memory_test.go (hunk 1 above)
Comment (evan2645) [reply pkg/server/plugin/ca/memory/memory_test.go]: Updated to use `assert.Error`

**Pull request merged**